### ISSUE TITLE
Subclassing Parameter

### ISSUE THREAD

**Issue by thomasdeneux** [holoviz/param#353]
Hello, 

i try to add my own information to `param.Parameter` instances by subclassing it

here is my attempt (the additional information that i want to store in parameters is whether they should remain enabled in the graphic interface when a run has started)

```
import param as pm

# Add run_enabled slot to pm.Parameter classes
class RunEnabledInfo:

    __slots__ = ['run_enabled']

    def __init__(self, *args, run_enabled=None, **kwargs):
        super(MyBoolean, self).__init__(*args, **kwargs)
        self.run_enabled = run_enabled

class MyBoolean(pm.Boolean, RunEnabledInfo):
    pass

class MyNumber(pm.Number, RunEnabledInfo):
    pass
```

But i get

```
Traceback (most recent call last):
  File "E:/Users/Thomas/Python/Naivia/AlphaI/test.py", line 18, in <module>
    class MyBoolean(pm.Boolean, RunEnabledInfo):
  File "C:\ProgramData\Anaconda3\lib\site-packages\param\parameterized.py", line 503, in __new__
    return type.__new__(mcs,classname,bases,classdict)
TypeError: multiple bases have instance lay-out conflict
```

The problem seems to be with `MyBoolean` inheriting from two classes. Indeed the following code works, but i wouldn't like to rewrite it for every possible parameter type! Any suggestion how to do?

```
class MyBoolean(pm.Boolean):

    __slots__ = ['run_enabled']

    def __init__(self, *args, run_enabled=None, **kwargs):
        super(MyBoolean, self).__init__(*args, **kwargs)
        self.run_enabled = run_enabled
```

**Comment by jbednar** [comment]
Can't you use the parameter's `precedence` value to determine if they should be shown, without having to do any subclassing?

**Comment by philippjfr** [comment]
>should remain enabled in the graphic interface when a run has started

In Panel setting ``parameter.constant=True`` does this.

**Comment by thomasdeneux** [comment]
@jbednar Sure i could use an available slot such as `precedence`, but i might need in the future to add more than one to my parameters, so this is why i try to subclass

@philippjfr I am not using Panel for my GUI, and in any case the `run_enabled` attribute i want to add should really be set at the parameter level and not at the GUI level, because it is needed in other places (for example, when there is no active run and a parameter is changed, to infer whether some reinitialization is needed before the next run)

**Comment by jbednar** [comment]
Subclassing many subclasses in a hierarchy like Param's is not likely to be a good solution to _any_ problem I can think of. :-)  If the existing support for indicating whether a Parameter should be displayed by a GUI isn't sufficient, you should be able to set an arbitrary attribute directly on the parameter instance, but (a) that will presumably cause each instance to get a dictionary (right now they should only have slots, which require less memory), and (b) you probably have to add that info *after* the Parameter is created, as I think it will error if you add it in the constructor.  So I'd hope that you can use the existing slots for this (precedence, constant, or readonly).  If not, maybe consider making a PR to add a `metadata` slot to all Parameters, though that has a (small) memory cost for all Param users while so far not being needed by anyone else...

**Comment by philippjfr** [comment]
>in any case the run_enabled attribute i want to add should really be set at the parameter level and not at the GUI level

I wasn't saying that it's set at the GUI level, just that if it is set the GUI will disable the corresponding widget.

**Comment by thomasdeneux** [comment]
ok @philippjfr i didn't understand exactly before; so to rephrase my own need, i would like when i define parameters to set an additional attribute that will tell which ones will need to be set constant during runs

@jbednar in general there are potentially many information that one might want to attach to parameters: another example is dependencies between parameters, i.e. some parameters need to be displayed in the GUI only if some other have a specific value, etc., but this seems even more complicate to set into place (for the moment this dependencies appear in a separate function that sets visibility of parameters in the GUI, but it would be nicer if they could appear directly when defining parameters)

in any case for the moment i will user the `precedence` slot to store information; i think that you can close this issue, thanks

**Comment by jbednar** [comment]
Dependencies between parameters are supported already using `@param.depends`.  I know you aren't using Panel, but at the moment [Panel's documentation](https://panel.pyviz.org) is probably the best guide to  how to do this.  I can't think of a particular example that shows it, but in Panel it should be straightforward to make widgets appear and disappear depending on the state of another widget, by defining methods depending on one Parameter that change the precedence of other Parameters.  You should be able to do something similar with any widget toolkit.

BTW, what GUI toolkit are you actually using?  I'd be happy to see e.g. ParamQt appear someday. :-)

**Comment by philippjfr** [comment]
Note that ``@param.depends`` is simply a helper around ``parameterized.param.watch`` which is what you probably should be using if you're writing GUI code and is (mostly) the API that Panel uses internally.

**Comment by thomasdeneux** [comment]
Hey, great to learn about `@param.depends`. 
Besides, i had a look at Panel some time ago, and it is precisely because it does not handle Qt that i stayed with my own solution, which is a quite simple one. Below is how it looks; sharing it? why not.

![controls](https://user-images.githubusercontent.com/6928710/62790862-4548bb80-bacc-11e9-8a6c-e6e81dac3e5f.png)

**Comment by thomasdeneux** [comment]
Also, unfortunately i found out about the `watch` mechanism only after i had written my own action registration mechanism in a subclass of `pm.Parameterized`, and these days i am precisely considering switching to `watch`.

**Comment by jbednar** [comment]
From that screenshot, it looks like you are doing things that are well within what we originally designed Param for. Originally we were using Tk, and we were only controlling simulations rather than both simulations or physical systems, but otherwise we had basically the same needs. Given that we'd love to see Qt support in general, I would strongly encourage you to try to use what Param offers without any subclassing, and letting us know anything that doesn't fit -- it's very likely that it would make sense to put into Param itself rather than a subclass.  

A subclass would be suitable if you wanted to do something weird or out of scope, but here I think your needs should fit in, and if they don't, we should make them fit in.  So if you're up for it, I think we're up to help give you some advice, and if something reusable like (part of Panel) but for Qt comes out of it, that would be fabulous. In general, Panel is a good design to go by; our earlier packages ParamTk, ParamNB, and ParamBokeh aren't how we would recommend designing Param+Qt support (as they are much more limited and ad-hoc).

**Comment by thomasdeneux** [comment]
@jbednar Hello! Sorry for the long interruption during my vacation and the busy period that followed. So ok, i will start by separating the machinery for generating controls as in the screenshot above from the rest of my code and submit it to you for approval of the general way of doing.

Regarding my need for additions to Param, it is essentially the possibility to define new parameter attributes. The list of attributes that one might want to add would be user-dependent, but it is probably enough to create a single new slot `user`. I can create a pull request at the same time i will do the above proposition for the Qt support.

**Comment by jbednar** [comment]
Thanks, though please delay adding `user` if your immediate needs can be met by one of the existing slots.  If there is a very strong case for `user` beyond those existing slots, sure, we can add them, but otherwise it's better to postpone it indefinitely while knowing it can be added if it is ever truly needed.

**Comment by thomasdeneux** [comment]
Hello @jbednar ! Step by step i have advanced on the "Qt support for param". See PR #366 which first includes a few bug fixes in param that were needed, then the ParamQt.py file (, then some minor fixes for Travis). This file contains an example in the `if __name__ == "__main__"` block, I suggest you try it and then we discuss.

I took inspiration from Panel, though probably not in-depth. I have some questions and requests regarding param, briefly:
- i used the `precedence` slot to gather custom information such as control style and user's own specification (the example shows how the user creates and uses a new 'expert' flag); i would still prefer that a `user` slot exists
- i need a new `visible` slot, because then the event machinery is used to trigger display changes when this property is modified
- you want to limit the number of slots to spare memory usage, but i have another related concern that every parameter is instantiated at least twice, one at the Parameterized class level and another at the object level, while often only those at the object level are used; or are there usage recommandations to avoid this?
- i wondered what is the best way to create nested parameters, you will tell me whether the way i did is appopriate (see e.g. `all_par.shape_par.pos_par` in the example)
- those are the most significant points that i can think of, there will be plenty others for sure as we will discuss; also of course the question: would you envision a new holoviz entry for this ParamQt? i would not have the time to maintain it well alone.

Best

**Comment by jbednar** [comment]
Thanks!  I'll take a look. I'll reply to a few things now, but the rest will need to wait until next week.

For 1, I'd definitely not want the precedence slot to be used in this way for real work, and so yes, a `user` or `metadata` or `custom` slot that defaults to None and can be set to a dictionary would be much better.  Here, it sounds like you're tagging parameters with a declaration that they should be shown only in "expert" mode?  For that specific purpose, precedence should already work as a single integer.  Just define categories like SIMPLE=0.4, EXPERT=0.6, etc., setting the default to SIMPLE and the display_threshold to 0.5, while adding precedence=EXPERT when you declare the parameter.  Then the EXPERT parameters will only be shown when the display_threshold is set to EXPERT.

For 2, I'm not sure what the intent of `visible` is; maybe you could do a short screensharing demo to explain how this is being used? I'd normally think of `precedence` as controlling visibility, but not in a way that would be used often in callbacks, as you seem to be indicating.  

For 3, I don't think there's a problem there unless I'm misunderstanding something.  A Python program can create an arbitrary number of objects, but surely the code will have a finite number of possible classes, so the amount of memory taken up by the class's copy of the Parameter object should be limited.  Having a copy at the class level allows users to set new default values that are then inherited by all subsequent objects of that type, which provides an important mechanism for users to configure behavior of the code. Having a copy at the object level is optional, but where a programmer decides to allows that, the given parameter can then be set on a per-instance basis, which is sometimes important.  It's true that the object-level copy then shadows the class copy, but that doesn't make the class copy redundant, because _subsequent_ new instances of that class will start from the class's copy of the Parameter, not the object's. This design makes it possible to concisely specify a default value for a long section of code, then change the value later for subsequent objects that get instantiated.

For 4, I'm not sure I understand what your intent was, but normally one would nest a ParameterizedObject in a Parameter, not what you did there.  See :"Parameter subobjects" in https://panel.pyviz.org/user_guide/Param.html .

For 5, as long as you could be involved long enough to get automated tests running, we can make there be a new holoviz project for ParamQt and perform the minimal maintenance needed to keep it running at a basic level. Anything beyond that would depend on your own future needs and that of any other users (and potentially us, if we start using it with Qt at some point, which is not something I current plan on).

**Comment by philippjfr** [comment]
To be clear, instance level parameters are a new concept and unless you are accessing them explicitly they are in fact not copied. There are APIs to avoid making instance level copies entirely but since we still don't have any docs those too aren't documented.

**Comment by thomasdeneux** [comment]
Hello,

Oh of course i forgot to put a screenshot. Here are a few ones, but i strongly encourage you to try it yourself.

Regarding the points:
1 - so we agree to add a `user` or `custom` or `metadata` slot (this is the order for my own preference). Note that in the example, some parameters are shown if "show expert parameters" is checked _and_ some condition is met on the value of other parameter. So the example user code already is quite complex with two fields `'show'` and `'active'` inside the `user` slot structure.

2 - screenshots below show how certain parameters are not shown depending on the value of other parameters; the user custom code mentioned just above controls the `visible` slot value, and this triggers an event that leads to the appropriate display update (note that control visibility depends both on the `visible` slot value and on the ). Alternatively user might want to keep the control visible but disable it using the `readonly` slot, this is also illustrated with the last screenshot.

3 - as you can see my usage of `param` is to create a unique instance that contains a lot of parameters, including nested Parameterized objects, and will be a global variable in my program; quite different from instantiating a number of objects from the same Parameterized subclass. I understand that i might consider not instantiating at all but use parameters at the class level directly. This is a bit unnatural to me as i tend to consider a parameters structure as an object instance, but why not; the present `ParamQt` would need to be adapted for such usage.

4 - oh now i better understand your way to automatically display subparameters specific to a certain choice of other parameters. I have reasons though to prefer a rigid large parameter structures and less small specific Parameterized class definitions, even though some parameters in some cases exist but are not used (hence should not made visible). Note that i defined subparameter structures as `param.Parameter`, but instead of setting the value in the default, I set it in the `__init__`  method: this allows my IDE to know that, for example, `all_par.shape_par` is an object of class `ShapeParam` and thus suggest the accurate parameter attributes.

5 - yes i will write tests, this could take me maybe another about two months? I am afraid however that tests are more difficult to write when there is some graphics involved... I'll seek inspiration again from `Panel`.

![Sans titre](https://user-images.githubusercontent.com/6928710/70983528-2fba5280-20b9-11ea-94f3-ff19fc077612.png)
![Sans titre](https://user-images.githubusercontent.com/6928710/70984169-4e6d1900-20ba-11ea-8288-9ae7d32ebeae.png)
![Sans titre](https://user-images.githubusercontent.com/6928710/70983851-b7a05c80-20b9-11ea-82df-008e0a9838c1.png)
![Sans titre](https://user-images.githubusercontent.com/6928710/70983984-f46c5380-20b9-11ea-81cf-3707c370d8ef.png)
![Sans titre](https://user-images.githubusercontent.com/6928710/70985890-30ed7e80-20bd-11ea-82b5-a44b5ca2a606.png)

**Comment by thomasdeneux** [comment]
Hello, i reactivate the discussion following some comments on PR #366.

I have written a Python Qt program (by the way it is out now, see https://learningrobots.co) which relies on `param` for storing its parameters. I have put a significant effort in separating the Qt interface for controlling the parameters from the rest of the program and pave the way to a new ParamQt package available to everybody. _To do it properly though i would need your feedback on what i have built so far, and advance the discussion with you on the particular question of nested parameters._

Indeed, in a desktop app it is often preferable to have all parameters contained in a single parameters structure rather than being distributed among the objects using them. This leads to a large `pm.Parameterized` object with nested `pm.Parameterized` object. The way i implemented can be seen in PR #366: contrarily to the first versions, i now use `pm.Parameterized` subclasses at the class level and do not instantiate them. The main advantage is to get elegant user code, with no need to instantiate nested parameters in the `__init__` method, here is a subsample:

```
   class PosPar(pm.Parameterized):

        shape = GObjectSelector('star',
                                ['circle', 'polygon', 'star'],
                                style='button',
                                doc="select drawing's shape",
                                allow_None=True)
		...

    class EdgePar(pm.Parameterized):

        color = GColor('#000000', allow_None=True)
		...

    class FillPar(pm.Parameterized):

        color = GColor('#008080', allow_None=True)
		...

    class ShapePar(pm.Parameterized):

        antialiasing = GBoolean(True, expert=True)

        pos = PosPar
        edge = EdgePar
        fill = FillPar
```

I have some problems though:
- the `@depends` decorator is not fully working, i will need to bring some corrections and/or additions to `param` (i do not remember the details of the problem i encountered, i can send a description after i fully get back to it)
- if class `ShapePar` above is instantiated, its `pos`, `edge`, `fill` subparameters are shared between instances, this is not the expected behavior, some consequent work might be needed in `param` if one wants to implement the expected behavior.

So, thank you for your comments on this question and more generally on this ParamQt contribution (for a fast lookaround, the example from which the screenshot above are taken executes when running ParamQt.py).

**Comment by tonyfast** [comment]
Closing, but feel free to open new issues for any specific requests.

**Comment by thomasdeneux** [comment]
@tonyfast @ceball @jbednar Hello, as you know i have developed a Qt interface for param. You are closing this thread, but for me the discussion is still open! So let me reactivate it once more. Also i just created a new repository for this code: [param_qt](https://github.com/thomasdeneux/param_qt), so you can close PR #366. I use this code daily, but i cannot make it public because some questions detailed above require answers:

- I had to hack `param` in some ways and would need to discuss additions to `param` that would unnecessitate these hacks.
- I need feedback from the param team on what was done so far!! And on how this is coherent with the param philosophy (one main question is how i handle nested parameters). In the absence of such feedback i keep the code only for myself, and there is a risk that over time i will adapt it too much to my own needs and make it less user friendly to others.
- I would need some advices on how to code tests for a graphic toolbox, and even better would be if someone thinks the Qt interface is a valuable project and can bring help especially with the tests.

So please someone have a look, test the program, sincerely this is not so long to do.

**Comment by jbednar** [comment]
Hi Thomas,

I hadn't seen your latest message appear above. It's great to see param_qt as a repository at long last!  And https://learningrobots.co looks really cool.  Yes, I do think a Qt interface is a valuable project, and I can easily imagine wanting to use it in our own projects at some point (just not anything we're currently contemplating).

Now that you've opened the new repo, let's separate this discussion between issues around the design of param_qt (which should be opened as issues at that new repo) and changes needed in Param (which should be opened here as specific individual PRs where feasible). I'll comment on the ShapePar issue on the new repo, and please open new issues or a PR on this repo for `visible` (or `user`, which would subsume it), `_batch_call_watchers`, and an other changes that are meant to be done at the Param level.

**Comment by thomasdeneux** [comment]
@jbednar Note that i was not able to make progress on [paramqt](https://github.com/thomasdeneux/paramqt) because i couldn't prevent interns of putting into it features that are too specific of our own application! I have kept postponing the work of disentangling our 'custom' part from the true 'toolbox' part; future will tell if i succeed doing it at some point. Please tell me if you ear about people interested by a Qt interface, that can increase motivation :)

There is also `pyqtgraph` [`ParameterTree`](https://pyqtgraph.readthedocs.io/en/latest/parametertree/index.html) implementation (example images: [1](https://stackoverflow.com/questions/27505371/subclassing-pyqtgraph-parametertree-issue), [2](https://github.com/pyqtgraph/pyqtgraph/issues/1314), [3](https://github.com/pyqtgraph/pyqtgraph/issues/1379)), which relies on its own, limited, format for parameters (see [example](https://github.com/pyqtgraph/pyqtgraph/blob/master/examples/_paramtreecfg.py)). Maybe would it make sense to connect `param` parameter representation with this `pyqtgraph` interface system, but that is also another project!

**Comment by jbednar** [comment]
Yes, the [pyqtgraph](https://github.com/pyqtgraph/pyqtgraph/issues/1379) stuff looks cool and very much on the same wavelength as Param/Panel.  I'd personally love to see ParamQT usable, but because we mainly hear from current users, we don't much hear about Qt support.  I'm sure if there was QT support people would use it!
